include/wil/safearrays.h

```diff
@@ -0,0 +1,889 @@
+        }
+        // Uses Inferred Type
+        template<typename T = element_t, typename wistd::enable_if<is_type_set<T>::value, int>::type = 0>
+        result create(UINT elements, LONG lowerBound = 0)
+        {
+            constexpr auto vt = static_cast<VARTYPE>(details::infer_safearray_type_traits<T>::vartype);
+            auto bounds = SAFEARRAYBOUND{ elements, lowerBound };
+            return err_policy::HResult(_create(vt, 1, &bounds));
+        }
+
+        //! Create a Copy
+        //! Replaces the current SAFEARRAY (if any) with a new one that is created by duplicating the elements in
+        //! the SAFEARRAY* provided as a parameter in psaSource
+        result create_copy(pointer psaSource)
+        {
+            WI_ASSERT(psaSource != nullptr);
+            return err_policy::HResult([&]()
+                {
+                    auto psa = storage_t::policy::invalid_value();
+                    RETURN_IF_FAILED(::SafeArrayCopy(psaSource, &psa));
+                    storage_t::reset(psa);
+                    return S_OK;
+                }());
+        }
+
+        //! @name Property Helpers
+        //! Functions to determine the size and shape of the SAFEARRAY, which
+        //! can have an arbitrary number of dimensions and each dimension can
+        //! it's own size and boundaries.
+        //! @{
+
+        //! Indicates the number of dimensions in the SAFEARRAY.  The 1st Dimension
+        //! is always 1, and there never is a Dimension 0.
+        ULONG dims() const WI_NOEXCEPT { return ::SafeArrayGetDim(storage_t::get()); }
+        //! Returns the lowest possible index for the given dimension.
+        result lbound(UINT dim, LONG* result) const
+        {
+            WI_ASSERT(result != nullptr);
+            WI_ASSERT((dim > 0) && (dim <= dims()));
+            return err_policy::HResult(::SafeArrayGetLBound(storage_t::get(), dim, result));
+        }
+        //! Returns the highest possible index for the given dimension.
+        result ubound(UINT dim, LONG* result) const
+        {
+            WI_ASSERT(result != nullptr);
+            WI_ASSERT((dim > 0) && (dim <= dims()));
+            return err_policy::HResult(::SafeArrayGetUBound(storage_t::get(), dim, result));
+        }
+        //! 1D Specialization of lbound
+        result lbound(LONG* result) const
+        {
+            WI_ASSERT(dims() == 1);
+            WI_ASSERT(result != nullptr);
+            return err_policy::HResult(::SafeArrayGetLBound(storage_t::get(), 1, result));
+        }
+        //! 1D Specialization of ubound
+        result ubound(LONG* result) const
+        {
+            WI_ASSERT(dims() == 1);
+            WI_ASSERT(result != nullptr);
+            return err_policy::HResult(::SafeArrayGetUBound(storage_t::get(), 1, result));
+        }
+        //! Indicates the total number of elements across all the dimensions
+        result count(ULONG* result) const

```

Comment:
`size` is more consistent with STL containers

Reply:
Some STL containers use `count`, but all of them use `size`, so I will change it.  I suppose "Count" came from the shim function "SafeArrayCountElements", so that will become "SafeArrayGetTotalSize".  I guess `dim_elements` should become `dim_size`, and it's shim should be "SafeArrayGetDimSize", which gives them a symmetry they didn't have before.